src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+            int l = -1; // known < key
+            int r = transitionRange(src, position);   // known > key
+            position += 2;
+
+            while (l + 1 < r)
+            {
+                int m = (l + r + 1) / 2;
+                int childTransition = src.get(position + m) & 0xFF;
+                int cmp = Integer.compare(key, childTransition);
+                if (cmp < 0)
+                    r = m;
+                else if (cmp > 0)
+                    l = m;
+                else
+                    return m;
+            }
+
+            return -r - 1;
+        }
+
+        @Override
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            assert childIndex >= 0;
+            int range = transitionRange(src, position);
+            assert childIndex < range;
+            return -readBytes(src, position + 2 + range + bytesPerPointer * childIndex);
+        }
+
+        @Override
+        public long greaterTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            if (searchIndex < 0)
+                searchIndex = -1 - searchIndex;
+            else
+                ++searchIndex;
+            if (searchIndex >= transitionRange(src, position))
+                return defaultValue;
+            return transition(src, position, positionLong, searchIndex);
+        }
+
+        public long lesserTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            if (searchIndex == 0 || searchIndex == -1)
+                return defaultValue;
+            if (searchIndex < 0)
+                searchIndex = -2 - searchIndex;
+            else
+                --searchIndex;
+            return transition(src, position, positionLong, searchIndex);
+        }
+
+        @Override
+        public int transitionByte(ByteBuffer src, int position, int childIndex)
+        {
+            return childIndex < transitionRange(src, position) ? src.get(position + 2 + childIndex) & 0xFF : Integer.MAX_VALUE;
+        }
+
+        @Override
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 2 + node.childCount() * (1 + bytesPerPointer);
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            int childCount = node.childCount();
+            assert childCount > 0;
+            assert childCount < 256;
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+            dest.writeByte(childCount);
+
+            for (int i = 0; i < childCount; ++i)
+                dest.writeByte(node.transition(i));
+            for (int i = 0; i < childCount; ++i)
+                writeBytes(dest, -node.serializedPositionDelta(i, nodePosition));
+        }
+    };
+
+    static final Sparse12 SPARSE_12 = new Sparse12();
```

Comment:
nit: Should probably move these into `Sparse12` itself to avoid potential classloading deadlocks